File: NC-150/07_Trees/60_serialize_and_deserialize_binary_tree.py
"""
LeetCode 297: Serialize and Deserialize Binary Tree
https://leetcode.com/problems/serialize-and-deserialize-binary-tree/

Design an algorithm to serialize and deserialize a binary tree. 
Serialization is the process of converting a data structure into a 
sequence of bits so that it can be stored or transmitted.

Time Complexity: O(n) for both serialize and deserialize
Space Complexity: O(n) for storing the tree representation
"""

from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Codec:
    """Solution using DFS Preorder traversal"""
    
    def serialize(self, root: Optional[TreeNode]) -> str:
        """Encodes a tree to a single string."""
        result = []
        
        def dfs(node):
            if not node:
                result.append("null")
                return
            
            result.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        
        dfs(root)
        return ",".join(result)

    def deserialize(self, data: str) -> Optional[TreeNode]:
        """Decodes your encoded data to tree."""
        values = iter(data.split(","))
        
        def dfs():
            val = next(values)
            if val == "null":
                return None
            
            node = TreeNode(int(val))
            node.left = dfs()
            node.right = dfs()
            return node
        
        return dfs()


class CodecBFS:
    """Alternative solution using BFS Level Order traversal"""
    
    def serialize(self, root: Optional[TreeNode]) -> str:
        """Encodes a tree to a single string."""
        if not root:
            return "null"
        
        result = []
        queue = deque([root])
        
        while queue:
            node = queue.popleft()
            
            if node:
                result.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
            else:
                result.append("null")
        
        return ",".join(result)

    def deserialize(self, data: str) -> Optional[TreeNode]:
        """Decodes your encoded data to tree."""
        if data == "null":
            return None
        
        values = data.split(",")
        root = TreeNode(int(values[0]))
        queue = deque([root])
        i = 1
        
        while queue and i < len(values):
            node = queue.popleft()
            
            # Left child
            if values[i] != "null":
                node.left = TreeNode(int(values[i]))
                queue.append(node.left)
            i += 1
            
            # Right child
            if i < len(values) and values[i] != "null":
                node.right = TreeNode(int(values[i]))
                queue.append(node.right)
            i += 1
        
        return root


def print_tree(root):
    """Helper to print tree level order"""
    if not root:
        print("null")
        return
    
    result = []
    queue = deque([root])
    
    while queue:
        node = queue.popleft()
        if node:
            result.append(str(node.val))
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append("null")
    
    # Remove trailing nulls
    while result and result[-1] == "null":
        result.pop()
    
    print(result)


# Test
if __name__ == "__main__":
    # Test case: [1,2,3,null,null,4,5]
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(5)
    
    print("=== DFS Codec ===")
    codec_dfs = Codec()
    
    print("Original tree: ", end="")
    print_tree(root)
    
    serialized = codec_dfs.serialize(root)
    print(f"Serialized: {serialized}")
    
    deserialized = codec_dfs.deserialize(serialized)
    print("Deserialized tree: ", end="")
    print_tree(deserialized)
    
    print("\n=== BFS Codec ===")
    codec_bfs = CodecBFS()
    
    serialized_bfs = codec_bfs.serialize(root)
    print(f"Serialized: {serialized_bfs}")
    
    deserialized_bfs = codec_bfs.deserialize(serialized_bfs)
    print("Deserialized tree: ", end="")
    print_tree(deserialized_bfs)

File: NC-150/07_Trees/60_serialize_and_deserialize_binary_tree.cpp
/*
 * LeetCode 297: Serialize and Deserialize Binary Tree
 * https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
 * 
 * Design an algorithm to serialize and deserialize a binary tree. 
 * Serialization is the process of converting a data structure into a 
 * sequence of bits so that it can be stored or transmitted.
 * 
 * Time Complexity: O(n) for both serialize and deserialize
 * Space Complexity: O(n) for storing the tree representation
 */

#include <iostream>
#include <string>
#include <sstream>
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Codec {
public:
    // Solution 1: Level Order (BFS) approach
    
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) return "null";
        
        string result;
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            
            if (node) {
                result += to_string(node->val) + ",";
                q.push(node->left);
                q.push(node->right);
            } else {
                result += "null,";
            }
        }
        
        // Remove trailing nulls for cleaner output
        return result;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data == "null" || data.empty()) return nullptr;
        
        stringstream ss(data);
        string token;
        
        getline(ss, token, ',');
        TreeNode* root = new TreeNode(stoi(token));
        
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty() && getline(ss, token, ',')) {
            TreeNode* node = q.front();
            q.pop();
            
            // Left child
            if (token != "null") {
                node->left = new TreeNode(stoi(token));
                q.push(node->left);
            }
            
            // Right child
            if (getline(ss, token, ',') && token != "null") {
                node->right = new TreeNode(stoi(token));
                q.push(node->right);
            }
        }
        
        return root;
    }
};

// Alternative: DFS Preorder approach
class CodecDFS {
public:
    string serialize(TreeNode* root) {
        string result;
        serializeHelper(root, result);
        return result;
    }
    
    TreeNode* deserialize(string data) {
        stringstream ss(data);
        return deserializeHelper(ss);
    }
    
private:
    void serializeHelper(TreeNode* node, string& result) {
        if (!node) {
            result += "null,";
            return;
        }
        
        result += to_string(node->val) + ",";
        serializeHelper(node->left, result);
        serializeHelper(node->right, result);
    }
    
    TreeNode* deserializeHelper(stringstream& ss) {
        string token;
        getline(ss, token, ',');
        
        if (token == "null" || token.empty()) return nullptr;
        
        TreeNode* node = new TreeNode(stoi(token));
        node->left = deserializeHelper(ss);
        node->right = deserializeHelper(ss);
        
        return node;
    }
};

// Helper to print tree (for verification)
void printTree(TreeNode* root) {
    if (!root) {
        cout << "null" << endl;
        return;
    }
    
    queue<TreeNode*> q;
    q.push(root);
    cout << "[";
    
    while (!q.empty()) {
        TreeNode* node = q.front();
        q.pop();
        
        if (node) {
            cout << node->val << ",";
            q.push(node->left);
            q.push(node->right);
        } else {
            cout << "null,";
        }
    }
    cout << "]" << endl;
}

int main() {
    // Test case: [1,2,3,null,null,4,5]
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->right->left = new TreeNode(4);
    root->right->right = new TreeNode(5);
    
    Codec codec;
    
    cout << "Original tree: ";
    printTree(root);
    
    string serialized = codec.serialize(root);
    cout << "Serialized: " << serialized << endl;
    
    TreeNode* deserialized = codec.deserialize(serialized);
    cout << "Deserialized tree: ";
    printTree(deserialized);
    
    return 0;
}

